src/ext/random.c

```diff
@@ -5,24 +5,63 @@
 #include <ext/standard/php_rand.h>
 
 #include "configuration.h"
+#include "ddtrace.h"
 #include "third-party/mt19937-64.h"
 
-void dd_trace_seed_prng(TSRMLS_D) {
+ZEND_EXTERN_MODULE_GLOBALS(ddtrace);
+
+void ddtrace_seed_prng(TSRMLS_D) {
     if (get_dd_trace_debug_prng_seed() > 0) {
         init_genrand64((unsigned long long)get_dd_trace_debug_prng_seed());
     } else {
         init_genrand64((unsigned long long)GENERATE_SEED());
     }
 }
 
-static long long generate_id() {
+void ddtrace_init_span_id_stack(TSRMLS_D) {
+    DDTRACE_G(root_span_id) = 0;
+    DDTRACE_G(span_ids_top) = NULL;
+}
+
+void ddtrace_free_span_id_stack(TSRMLS_D) {
+    while (DDTRACE_G(span_ids_top) != NULL) {
+        ddtrace_span_ids_t *stack = DDTRACE_G(span_ids_top);
+        DDTRACE_G(span_ids_top) = stack->next;
+        efree(stack);
+    }
+}
+
+uint64_t ddtrace_push_span_id(TSRMLS_D) {
+    ddtrace_span_ids_t *stack = ecalloc(1, sizeof(ddtrace_span_ids_t));
     // We shift one bit to get 63-bit
-    return (long long)(genrand64_int64() >> 1);
+    stack->id = (uint64_t)(genrand64_int64() >> 1);
```

Comment:
Looking at this function again in relation to another previouse comment. Seems to me that we need to reroll or handle the case when genrand64_int64 returns 0. 

Since `0` is handled as a special value elsewhere we need to to at least add 1 or something along the lines of 

```c
id = (uint64_t)((genrand64_int64()-1) >> 1) + 1);
```

To avoid those one in trillion chances of something going wrong :D

Reply:
Brilliant! 🔝